segfault on irods reconnect (4.4.0rc1)

Hi again,

I'm running `parrot_run nginx ...` to serve files from an iRODS 3.3.1 server.  At first it works, but then after the idle timeout I get a segfault when I try to fetch again.  (With 4.3.3 I also got a segfault, though I think this was before reconnect was fixed.)

I built with `./configure --debug --prefix ~/build --with-irods-path /usr/local/iRODS --without-system-{sand,allpairs,wavefront,makeflow,resource_monitor{,_visualizer},doc}` on an Ubuntu Precise x86_64 machine, and I set `PARROT_DEBUG_FLAGS="fatal error notice debug  channel process pstree  irods  remote  time pid"` .

nginx is configured with `events { use select; }` to prevent it trying to use the default but unsupported epoll.

I'll start digging, this is what I have so far:

```
[...working fine with many HTTP GET requests...]
2015/04/13 11:52:22.64 parrot_run[19460] <child:19466> irods: rcDataObjOpen irods-cgp-dev:1247 /cgp-dev/intproj/_live_tmp/nfs/cancer_trk0006/00000020/172918.v1.wt.pindel.bw
2015/04/13 11:52:22.65 parrot_run[19460] <child:19466> irods: = 3
2015/04/13 11:52:22.65 parrot_run[19460] <child:19466> irods: rcObjStat irods-cgp-dev:1247 /cgp-dev/intproj/_live_tmp/nfs/cancer_trk0006/00000020/172918.v1.wt.pindel.bw
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: = 1
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: rcDataObjLseek irods-cgp-dev:1247 3 1900544
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: = 0
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: rcDataObjRead irods-cgp-dev:1247 3 32768
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: = 32768
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: rcDataObjClose irods-cgp-dev:1247 3
2015/04/13 11:52:22.66 parrot_run[19460] <child:19466> irods: = 0
# I wait, then try a HTTP HEAD request...
2015/04/13 12:01:02.22 parrot_run[19460] debug: poll: got event POLLIN on fd 12
2015/04/13 12:01:02.22 parrot_run[19460] process: pid 19466 woken from wait state
2015/04/13 12:01:02.23 parrot_run[19460] <child:19466> irods: discarding stale connection to irods-cgp-dev:1247
2015/04/13 12:01:02.26 parrot_run[19460] <child:19466> irods: connecting to irods-cgp-dev:1247
2015/04/13 12:01:02.43 parrot_run[19460] <child:19466> irods: logging in...
2015/04/13 12:01:02.43 parrot_run[19460] <child:19466> irods: rcDataObjOpen irods-cgp-dev:1247 /cgp-dev/intproj/_live_tmp/nfs/cancer_trk0005/00000018/955633.bam
2015/04/13 12:01:02.71 parrot_run[19460] <child:19466> irods: = 3
2015/04/13 12:01:02.71 parrot_run[19460] <child:19466> irods: rcObjStat irods-cgp-dev:1247 /cgp-dev/intproj/_live_tmp/nfs/cancer_trk0005/00000018/955633.bam
2015/04/13 12:01:02.71 parrot_run[19460] <child:19466> notice: received signal 11 (Segmentation fault), killing all my children...
2015/04/13 12:01:02.71 parrot_run[19460] <child:19466> process: killing pid 19465
2015/04/13 12:01:02.71 parrot_run[19460] <child:19466> process: killing pid 19466
2015/04/13 12:01:02.71 parrot_run[19460] <child:19466> notice: sending myself 11 (Segmentation fault), goodbye!
```
and
```
[3]+  Segmentation fault      (core dumped) ./nginx-parrot.sh
mca@cgp-1-1-02:~/gitwk-cgp/nginx-mca$ gdb ~/build/bin/parrot_run core
GNU gdb (Ubuntu/Linaro 7.4-2012.04-0ubuntu2.1) 7.4-2012.04
Copyright (C) 2012 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
For bug reporting instructions, please see:
<http://bugs.launchpad.net/gdb-linaro/>...
Reading symbols from /nfs/users/nfs_m/mca/build/bin/parrot_run...done.
[New LWP 19460]

warning: Can't read pathname for load map: Input/output error.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/nfs/users/nfs_m/mca/build/bin/parrot_run /nfs/users/nfs_m/mca/build/sbin/nginx'.
Program terminated with signal 11, Segmentation fault.
#0  0x00007f510fbf63b7 in kill () at ../sysdeps/unix/syscall-template.S:82
82      ../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) bt
#0  0x00007f510fbf63b7 in kill () at ../sysdeps/unix/syscall-template.S:82
#1  0x0000000000423b96 in pfs_process_kill_everyone (sig=11) at pfs_process.cc:59
#2  <signal handler called>
#3  0x00000000004a64b0 in procApiRequest (conn=0x73646f7200000000, apiNumber=633, inputStruct=0x7fff1a5ad590, inputBsBBuf=0x0, outStruct=0x7fff1a5ac578, outBsBBuf=0x0) at /usr/local/iRODS/lib/core/src/procApiRequest.c:46
#4  0x000000000049d3c7 in rcObjStat (conn=0x73646f7200000000, dataObjInp=0x7fff1a5ad590, rodsObjStatOut=0x7fff1a5ac578) at /usr/local/iRODS/lib/api/src/rcObjStat.c:72
#5  0x0000000000447a62 in irods_reli_stat (host=0x100c4d8 "irods-cgp-dev:1247", path=0x7fff1a5ac590 "/cgp-dev/intproj/_live_tmp/nfs/cancer_trk0005/00000018/955633.bam", info=0x7fff1a5adbb0) at irods_reli.cc:437
#6  0x0000000000446441 in pfs_service_irods::stat (this=0x7174a0, name=0x100b4d0, info=0x7fff1a5adbb0) at pfs_service_irods.cc:124
#7  0x000000000044615b in pfs_file_irods::fstat (this=0x100b4c0, info=0x7fff1a5adbb0) at pfs_service_irods.cc:50
#8  0x000000000042fe74 in pfs_table::fstat (this=0xdf3e30, fd=9, b=0x7fff1a5adbb0) at pfs_table.cc:1018
#9  0x000000000042643b in pfs_fstat (fd=9, buf=0x7fff1a5adbb0) at pfs_sys.cc:169
#10 0x000000000041a33f in decode_stat (p=0xdeada0, entering=1, syscall=5, args=0xdece18) at pfs_dispatch64.cc:440
#11 0x000000000041e002 in decode_syscall (p=0xdeada0, entering=1) at pfs_dispatch64.cc:1438
#12 0x0000000000423a1e in pfs_dispatch64 (p=0xdeada0, signum=0) at pfs_dispatch64.cc:2979
#13 0x000000000041938f in pfs_dispatch (p=0xdeada0, signum=0) at pfs_dispatch.cc:3221
#14 0x00000000004093ba in handle_event (pid=19466, status=34175, usage=0x7fff1a5cf8b0) at parrot_run.cc:338
#15 0x000000000040b340 in main (argc=6, argv=0x7fff1a5d02c8) at parrot_run.cc:1014
(gdb) quit
```

* mcast/cctools@c41de24e isn't building for me at the moment (below), I can poke it more tomorrow.
* The `... --with-irods-path /usr/local/iRODS` looks like mcast/irods-legacy@7dfcaab.
  * It seems to be vanilla 3.3.1, presumably from a tarball because there are no vc files, plus some small tweaks and built files.  For now my production servers are tied to the v3 series.
  * The upstream https://github.com/irods/irods doesn't (AIUI) go back as far as v3 so there isn't a nice way to get the union of all commits in the irods history.
  * I was planning to instrument this and rebuild with cctools-4.4.0rc1 to get more idea what's happening.

Details...
```
$ make install
[...many files compiled...]
g++ -o pfs_service_local.o -c -I/nfs/users/nfs_m/mca/build/src/cctools/dttools/src -I/nfs/users/nfs_m/mca/build/src/cctools/work_queue/src -I/nfs/users/nfs_m/mca/build/src/cctools/ftp_lite/src -I/nfs/users/nfs_m/mca/build/src/cctools/parrot/src -I/nfs/users/nfs_m/mca/build/src/cctools/chirp/src -D__EXTENSIONS__ -D_LARGEFILE64_SOURCE -D__LARGE64_FILES -Wall -Wextra -Werror -Wno-unused-parameter -Wno-unknown-pragmas -Wno-deprecated-declarations -Wno-unused-const-variable -fPIC -DHAS_IRODS -DIRODS_USES_PLUGINS -DIRODS_USES_HPP -DHAS_LIBREADLINE -DHAS_ISNAN -DHAVE_ISNAN -DSQLITE_HAVE_ISNAN -DHAS_PREAD -DHAS_PWRITE -DHAS_STRSIGNAL -DHAS_USLEEP -DHAVE_USLEEP -DHAS_UTIME -DHAVE_UTIME -DHAS_FTS_H -DHAS_INTTYPES_H -DHAVE_INTTYPES_H -DHAS_STDINT_H -DHAVE_STDINT_H -DHAS_SYS_STATFS_H -DHAS_SYS_STATVFS_H -DHAS_SYS_XATTR_H -DHAS_SYSLOG_H -DINSTALL_PATH='"/nfs/users/nfs_m/mca/build"' -g -D_REENTRANT -D_GNU_SOURCE -DBUILD_USER='"mca"' -DBUILD_HOST='"cgpfoo"' -DCCTOOLS_VERSION_MAJOR=4 -DCCTOOLS_VERSION_MINOR=4 -DCCTOOLS_VERSION_MICRO='"0"' -DCCTOOLS_VERSION='"4.4.0-TRUNK"' -DCCTOOLS_RELEASE_DATE='"04/13/2015"' -DCCTOOLS_CONFIGURE_ARGUMENTS='"--debug --prefix /nfs/users/nfs_m/mca/build --with-irods-path /nfs/users/nfs_m/mca/build/src/irods-legacy --without-system-sand --without-system-allpairs --without-system-wavefront --without-system-makeflow --without-system-resource_monitor --without-system-doc"' -DCCTOOLS_SYSTEM_INFORMATION='"Linux cgpfoo 2.6.32-44-server #98-Ubuntu SMP Mon Sep 24 17:41:33 UTC 2012 x86_64 x86_64 x86_64 GNU/Linux"' -DCCTOOLS_OPSYS_LINUX -DCCTOOLS_CPU_X86_64  pfs_service_local.cc
pfs_service_local.cc: In member function ‘virtual int pfs_file_local::canbenative(char*, size_t)’:
pfs_service_local.cc:110:33: error: format not a string literal and no format arguments [-Werror=format-security]
pfs_service_local.cc:110:33: error: format not a string literal and no format arguments [-Werror=format-security]
At global scope:
cc1plus: error: unrecognised command line option "-Wno-unused-const-variable" [-Werror]
cc1plus: all warnings being treated as errors
make[2]: *** [pfs_service_local.o] Error 1
make[2]: Leaving directory `/nfs/users/nfs_m/mca/build/src/cctools/parrot/src'
make[1]: *** [src] Error 2
make[1]: Leaving directory `/nfs/users/nfs_m/mca/build/src/cctools/parrot'
make: *** [parrot] Error 2
$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro 4.6.3-1ubuntu5' --with-bugurl=file:///usr/share/doc/gcc-4.6/README.Bugs --enable-languages=c,c++,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.6 --enable-shared --enable-linker-build-id --with-system-zlib --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.6 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --enable-plugin --enable-objc-gc --disable-werror --with-arch-32=i686 --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.6.3 (Ubuntu/Linaro 4.6.3-1ubuntu5) 
```

--- a/parrot/src/irods_reli.cc
+++ b/parrot/src/irods_reli.cc
@@ -127,6 +127,7 @@ static struct irods_server * connect_to_host( const char *hostport )
 			rcDisconnect(server->conn);
 			free(server);
 			server = 0;
+			hash_table_remove(connect_cache, hostport);
 		} else {
 			server->lastused = current;
 			return server;

